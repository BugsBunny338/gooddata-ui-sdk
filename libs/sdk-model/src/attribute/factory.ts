// (C) 2019 GoodData Corporation
import identity = require("lodash/identity");
import { IAttribute } from ".";

/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute}.
 *
 * @public
 */
export class AttributeBuilder implements IAttribute {
    public attribute: IAttribute["attribute"];

    /**
     * @internal
     */
    constructor(displayFormId: string) {
        this.attribute = {
            displayForm: { identifier: displayFormId },
            localIdentifier: `a_${displayFormId}`,
        };
    }

    public alias = (alias: string) => {
        this.attribute.alias = alias;
        return this;
    };

    public localId = (localId: string) => {
        this.attribute.localIdentifier = localId;
        return this;
    };

    public build = () => {
        return { attribute: this.attribute };
    };
}

/**
 * Function that will be called to perform modifications of an attribute before it is fully constructed.
 *
 * @public
 */
export type AttributeModifications = (builder: AttributeBuilder) => AttributeBuilder;

/**
 * Creates a new attribute with the specified display form identifier and optional modifications and localIdentifier.
 * @param displayFormId - identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(
    displayFormId: string,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(displayFormId);
    return modifications(builder).build();
}
