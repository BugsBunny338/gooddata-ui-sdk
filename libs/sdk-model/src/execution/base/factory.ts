// (C) 2020 GoodData Corporation
import { Identifier, Uri, IdentifierRef, UriRef, LocalIdRef } from ".";

/**
 * Creates an IdentifierRef from an identifier
 * @param identifier - identifier to use
 * @public
 */
export function idRef(identifier: Identifier): IdentifierRef {
    return { identifier };
}

/**
 * Creates an UriRef from an URI
 * @param uri - URI to use
 * @public
 */
export function uriRef(uri: Uri): UriRef {
    return { uri };
}

/**
 * Creates an LocalIdRef from a local identifier
 * @param localIdentifier - local identifier to use
 * @public
 */
export function localIdRef(localIdentifier: Identifier): LocalIdRef {
    return { localIdentifier };
}
