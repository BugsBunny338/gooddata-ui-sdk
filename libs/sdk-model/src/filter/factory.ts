// (C) 2019 GoodData Corporation
import {
    AttributeElements,
    IAbsoluteDateFilter,
    INegativeAttributeFilter,
    IPositiveAttributeFilter,
    IRelativeDateFilter,
} from ".";
import { attributeIdentifier, IAttribute } from "../attribute";

/**
 * Creates a new positive attribute filter.
 *
 * @param attributeOrId - either instance of attribute to create filter for or identifier of attribute's display form
 *  if the input is attribute object, then it is expected that attribute references display form by identifier
 * @param inValues - values to filter for; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newPositiveAttributeFilter(
    attributeOrId: IAttribute | string,
    inValues: AttributeElements | string[],
): IPositiveAttributeFilter {
    const displayFormId =
        typeof attributeOrId === "string" ? attributeOrId : attributeIdentifier(attributeOrId)!;
    const inObject: AttributeElements = Array.isArray(inValues) ? { values: inValues } : inValues;

    return {
        positiveAttributeFilter: {
            displayForm: { identifier: displayFormId },
            in: inObject,
        },
    };
}

/**
 * Creates a new negative attribute filter.
 *
 * @param attributeOrId - either instance of attribute to create filter for or identifier of attribute's display form;
 *  if the input is attribute object, then it is expected that attribute references display form by identifier
 * @param notInValues - values to filter out; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newNegativeAttributeFilter(
    attributeOrId: IAttribute | string,
    notInValues: AttributeElements | string[],
): INegativeAttributeFilter {
    const displayFormId =
        typeof attributeOrId === "string" ? attributeOrId : attributeIdentifier(attributeOrId)!;
    const notInObject: AttributeElements = Array.isArray(notInValues) ? { values: notInValues } : notInValues;

    return {
        negativeAttributeFilter: {
            displayForm: { identifier: displayFormId },
            notIn: notInObject,
        },
    };
}

/**
 * Creates a new absolute date filter.
 *
 * @param dateDataSetId - identifier of the date data set to filter on
 * @param from - start of the interval in ISO-8601 calendar date format
 * @param to - end of the interval in ISO-8601 calendar date format
 * @public
 */
export function newAbsoluteDateFilter(dateDataSetId: string, from: string, to: string): IAbsoluteDateFilter {
    return {
        absoluteDateFilter: {
            dataSet: { identifier: dateDataSetId },
            from,
            to,
        },
    };
}

/**
 * Defines date data set granularities that can be used in relative date filter.
 *
 * @public
 */
export const DateGranularity = {
    date: "GDC.time.date",
    week: "GDC.time.week_us",
    month: "GDC.time.month",
    quarter: "GDC.time.quarter",
    year: "GDC.time.year",
};

/**
 * Creates a new relative date filter.
 *
 * @param dateDataSetId - identifier of the date data set to filter on
 * @param granularity - granularity of the filters (month, year, etc.)
 * @param from - start of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @param to - end of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @public
 */
export function newRelativeDateFilter(
    dateDataSetId: string,
    granularity: string,
    from: number,
    to: number,
): IRelativeDateFilter {
    return {
        relativeDateFilter: {
            dataSet: { identifier: dateDataSetId },
            granularity,
            from,
            to,
        },
    };
}
