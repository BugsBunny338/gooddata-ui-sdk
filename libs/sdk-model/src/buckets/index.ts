// (C) 2019 GoodData Corporation
import isEmpty = require("lodash/isEmpty");
import { anyAttribute, AttributePredicate, IAttribute, idMatchAttribute, isAttribute } from "../attribute";
import { Identifier } from "../base";
import {
    anyMeasure,
    idMatchMeasure,
    IMeasure,
    isMeasure,
    measureDisableComputeRatio,
    MeasurePredicate,
} from "../measure";
import { isTotal, ITotal } from "../base/totals";

/**
 * TODO: SDK8: Add docs
 * TODO: SDK8: rename this; perhaps Item was the right name all along? :)
 *
 * @public
 */
export type AttributeOrMeasure = IMeasure | IAttribute;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface IBucket {
    localIdentifier?: Identifier;
    items: AttributeOrMeasure[];
    totals?: ITotal[];
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type BucketPredicate = (bucket: IBucket) => boolean;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export const anyBucket: BucketPredicate = _ => true;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export const idMatchBucket: (id: string) => BucketPredicate = id => bucket => bucket.localIdentifier === id;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type AttributeInBucket = {
    bucket: IBucket;
    idx: number;
    attribute: IAttribute;
};

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type MeasureInBucket = {
    bucket: IBucket;
    idx: number;
    measure: IMeasure;
};

//
// Type guards
//

/**
 * Type-guard testing whether the provided object is an instance of {@link IBucket}.
 *
 * @param obj - object to test
 * @public
 */
export function isBucket(obj: any): obj is IBucket {
    return !isEmpty(obj) && (obj as IBucket).items !== undefined;
}

//
// Functions
//

/**
 * Creates a new bucket with the provided id and all the specified content.
 *
 * @param id - bucket identifier
 * @param content - items to put into the bucket; attributes, measures and/or totals
 * @returns always new instance
 * @public
 */
export function newBucket(id: string, ...content: Array<AttributeOrMeasure | ITotal | undefined>): IBucket {
    const items: AttributeOrMeasure[] = [];
    const totals: ITotal[] = [];

    (content ? content : []).forEach(i => {
        if (!i) {
            return;
        }
        if (isAttribute(i) || isMeasure(i)) {
            items.push(i);
        } else if (isTotal(i)) {
            totals.push(i);
        } else {
            // TODO: SDK8: switch to invariant
            throw new Error("...");
        }
    });

    return {
        localIdentifier: id,
        items,
        totals,
    };
}

/**
 * Tests whether the provided bucket is empty = contains no items and no totals.
 *
 * @param bucket - bucket to test
 * @returns true if empty, false if not
 * @public
 */
export function bucketIsEmpty(bucket: IBucket): boolean {
    return !bucket || (bucket.items.length === 0 && (!bucket.totals || bucket.totals.length === 0));
}

/**
 * Gets first attribute matching the provided predicate from the bucket.
 *
 * If no predicate is provided, then the function defaults to anyAttribute predicate - meaning first found attribute
 * will be returned.
 *
 * This function also provides convenience to find attribute by its local identifier - if you pass predicate as
 * string the function will automatically create idMatchAttribute predicate.
 *
 * @param bucket - bucket to to search in
 * @param idOrFun - attribute identifier or instance of AttributePredicate; {@link anyAttribute} predicate is default
 * @returns undefined if no matching attribute is found
 * @public
 */
export function bucketAttribute(
    bucket: IBucket,
    idOrFun: string | AttributePredicate = anyAttribute,
): IAttribute | undefined {
    if (!bucket) {
        return;
    }

    const predicate = typeof idOrFun === "string" ? idMatchAttribute(idOrFun) : idOrFun;
    const compositeGuard = (obj: any): obj is IAttribute => {
        return isAttribute(obj) && predicate(obj);
    };

    return bucket.items.find(compositeGuard);
}

/**
 * Gets all attributes matching the provided predicate from the bucket.
 *
 * If no predicate is provided, then the function defaults to anyAttribute predicate - meaning all attributes
 * from the bucket will be returned.
 *
 * @param bucket - bucket to work with
 * @param predicate - attribute predicate; {@link anyAttribute} predicate is default
 * @returns empty list if none match
 * @public
 */
export function bucketAttributes(
    bucket: IBucket,
    predicate: AttributePredicate = anyAttribute,
): IAttribute[] {
    if (!bucket) {
        return [];
    }

    // need custom type-guard so as not to break type inference in filter() method
    const compositeGuard = (obj: any): obj is IAttribute => {
        return isAttribute(obj) && predicate(obj);
    };

    return bucket.items.filter(compositeGuard);
}

/**
 * Gets all measures matching the provided predicate from the bucket.
 *
 * If no predicate is provided, then the function defaults to anyMeasure predicate - meaning all measures from
 * the bucket will be returned.
 *
 * @param bucket - bucket to work with
 * @param predicate - measure predicate; {@link anyMeasure} predicate is default
 * @returns empty list if none match
 * @public
 */
export function bucketMeasures(bucket: IBucket, predicate: MeasurePredicate = anyMeasure): IMeasure[] {
    if (!bucket) {
        return [];
    }

    // need custom type-guard so as not to break type inference in filter() method
    const compositeGuard = (obj: any): obj is IMeasure => {
        return isMeasure(obj) && predicate(obj);
    };

    return bucket.items.filter(compositeGuard);
}

/**
 * Gets all attributes and measures from the bucket.
 *
 * @param bucket - bucket to work with
 * @returns empty list if no items
 * @public
 */
export function bucketItems(bucket: IBucket): AttributeOrMeasure[] {
    if (!bucket) {
        return [];
    }

    return bucket.items;
}

/**
 * Gets all totals from the bucket
 *
 * @param bucket - bucket to work with
 * @returns empty list if no totals
 * @public
 */
export function bucketTotals(bucket: IBucket): ITotal[] {
    if (!bucket) {
        return [];
    }

    return bucket.totals ? bucket.totals : [];
}

//
// Functions working with array of bucket
//

/**
 * Gets all attributes matching the provided predicate from a list of buckets.
 *
 * If no predicate is provided, then the function defaults to {@link anyAttribute} predicate - meaning all
 * attributes will be returned.
 *
 * @param buckets - list of buckets to get attributes from
 * @param predicate - attribute predicate; {@link anyAttribute} is default
 * @returns empty list if none match
 * @public
 */
export function bucketsAttributes(
    buckets: IBucket[],
    predicate: AttributePredicate = anyAttribute,
): IAttribute[] {
    if (!buckets || !buckets.length) {
        return [];
    }

    return buckets.map(b => bucketAttributes(b, predicate)).reduce((acc, items) => acc.concat(items), []);
}

/**
 * Gets all measures matching the provided predicate from a list of buckets.
 *
 * If no predicate is provided, then the function defaults to {@link anyMeasure} predicate - meaning all
 * measures will be returned.
 *
 * @param buckets - list of buckets to get measures from
 * @param predicate - measure predicate; {@link anyMeasure} is default
 * @returns empty list if none match
 * @public
 */
export function bucketsMeasures(buckets: IBucket[], predicate: MeasurePredicate = anyMeasure): IMeasure[] {
    if (!buckets || !buckets.length) {
        return [];
    }

    return buckets.map(b => bucketMeasures(b, predicate)).reduce((acc, items) => acc.concat(items), []);
}

/**
 * Finds bucket matching the provided predicate in a list of buckets.
 *
 * If no predicate is provided, then the function defaults to {@link anyBucket} predicate - meaning first
 * bucket in the list will be returned.
 *
 * This function also provides convenience to find bucket by local identifier - if you pass predicate as
 * string the function will automatically create idMatchBucket predicate.
 *
 * @param buckets - list of buckets to search
 * @param idOrFun - bucket predicate or string to match bucket by local identifier; {@link anyBucket} is default
 * @public
 */
export function bucketsFind(
    buckets: IBucket[],
    idOrFun: string | BucketPredicate = anyBucket,
): IBucket | undefined {
    const predicate = typeof idOrFun === "string" ? idMatchBucket(idOrFun) : idOrFun;

    return buckets.find(predicate);
}

/**
 * Finds attribute matching the provided predicate in a list of buckets. If found, the function returns an object
 * that contains bucket where the matched attribute is stored, index within that bucket and the attribute itself.
 *
 * This function also provides convenience to find attribute by local identifier - if you pass predicate as
 * string the function will automatically create idMatchAttribute predicate.
 *
 * @remarks See {@link AttributeInBucket}
 *
 * @param buckets - list of buckets to search
 * @param idOrFun - attribute predicate or string to find attribute by local identifier; no default
 * @returns first-found attribute matching the predicate, undefined if none match
 * @public
 */
export function bucketsFindAttribute(
    buckets: IBucket[],
    idOrFun: string | AttributePredicate,
): AttributeInBucket | undefined {
    if (!buckets || !buckets.length) {
        return;
    }

    const predicate = typeof idOrFun === "string" ? idMatchAttribute(idOrFun) : idOrFun;
    const typeAgnosticPredicate = (obj: any): boolean => {
        return isAttribute(obj) && predicate(obj);
    };

    for (const bucket of buckets) {
        const idx = bucket.items.findIndex(typeAgnosticPredicate);

        if (idx >= 0) {
            const item = bucket.items[idx];
            return isAttribute(item) ? { bucket, idx, attribute: item } : undefined;
        }
    }

    return undefined;
}

/**
 * Finds measure matching the provided predicate in a list of buckets. If found, the function returns an object
 * that contains bucket where the matched measure is stored, index within that bucket and the measure itself.
 *
 * This function also provides convenience to find measure by local identifier - if you pass predicate as
 * string the function will automatically create idMatchMeasure predicate.
 *
 * @remarks See {@link MeasureInBucket}
 *
 * @param buckets - list of buckets to search
 * @param idOrFun - measure predicate or string to find measure by local identifier; no default
 * @returns first-found measure matching the predicate, undefined if none match
 * @public
 */
export function bucketsFindMeasure(
    buckets: IBucket[],
    idOrFun: string | MeasurePredicate,
): MeasureInBucket | undefined {
    if (!buckets || !buckets.length) {
        return;
    }

    const predicate = typeof idOrFun === "string" ? idMatchMeasure(idOrFun) : idOrFun;
    const typeAgnosticPredicate = (obj: any): boolean => {
        return isMeasure(obj) && predicate(obj);
    };

    for (const bucket of buckets) {
        const idx = bucket.items.findIndex(typeAgnosticPredicate);

        if (idx >= 0) {
            const item = bucket.items[idx];
            return isMeasure(item) ? { bucket, idx, measure: item } : undefined;
        }
    }

    return undefined;
}

/**
 * Gets buckets with the provided local identifiers from a list of buckets.
 *
 * @param buckets - list of buckets to filter
 * @param ids - bucket identifiers
 * @returns empty list if none match
 * @public
 */
export function bucketsById(buckets: IBucket[], ...ids: string[]): IBucket[] {
    if (!ids || !ids.length) {
        return [];
    }

    return buckets.filter(b => b.localIdentifier && ids.indexOf(b.localIdentifier) >= 0);
}

/**
 * Gets all attributes and measures from a list of buckets.
 *
 * @param buckets - buckets to work with
 * @returns empty list if none
 * @public
 */
export function bucketsItems(buckets: IBucket[]): AttributeOrMeasure[] {
    return buckets.reduce((acc, b) => acc.concat(b.items), [] as AttributeOrMeasure[]);
}

/**
 * Gets all totals from a list of buckets
 *
 * @param buckets - buckets to work with
 * @returns empty list if none
 * @public
 */
export function bucketsTotals(buckets: IBucket[]): ITotal[] {
    return buckets.reduce((acc, b) => acc.concat(bucketTotals(b)), [] as ITotal[]);
}

/**
 * Tests whether all buckets in a list are empty (meaning neither has any items or totals defined)
 *
 * @param buckets - buckets to work with
 * @returns true if empty, false if not
 * @public
 */
export function bucketsIsEmpty(buckets: IBucket[]): boolean {
    if (!buckets || !buckets.length) {
        return true;
    }

    return buckets.every(bucketIsEmpty);
}

/**
 * Defines possible compute ratio sanitization rules.
 *
 * @public
 */
export enum ComputeRatioRule {
    /**
     * Compute ratio must not be used in any measure
     */
    NEVER,

    /**
     * Compute ratio can be used if there is just a single measure
     */
    SINGLE_MEASURE_ONLY,

    /**
     * Compute ratio can be used on any measure
     */
    ANY_MEASURE,
}

/**
 * Applies compute ratio rule to all measures in a list - this MAY be done to sanitize measure definitions
 * so that the computed results make sense when visualized in a chart
 *
 * The function will return a new list with updated measures according to the specified rule; see {@link ComputeRatioRule}.
 *
 * For convenience this function can work with list of measures AND attributes; attributes will be ignored
 * in processing and kept in resulting array as-is.
 *
 * @param items - list of attributes or measures to sanitize; attributes will be lef
 * @param rule - rule to apply; see {@link ComputeRatioRule}
 * @returns new list with modified measures; the original list and measures in it are left intact
 * @public
 */
export function computeRatioRules<T extends AttributeOrMeasure>(
    items: T[],
    rule: ComputeRatioRule = ComputeRatioRule.SINGLE_MEASURE_ONLY,
): T[] {
    if (!items) {
        return [];
    }

    if (rule === ComputeRatioRule.ANY_MEASURE) {
        return items;
    }

    if (items.length > 1 || rule === ComputeRatioRule.NEVER) {
        return items.map(disableComputeRatio);
    }

    return items;
}

function disableComputeRatio<T extends AttributeOrMeasure>(item: T): T {
    if (isMeasure(item)) {
        return measureDisableComputeRatio(item) as T;
    }
    return item;
}
