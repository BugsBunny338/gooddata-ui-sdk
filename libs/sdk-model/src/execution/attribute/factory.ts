// (C) 2019-2020 GoodData Corporation
import identity = require("lodash/identity");
import cloneDeep = require("lodash/cloneDeep");
import isEmpty = require("lodash/isEmpty");
import { IAttribute, isAttribute } from "./index";
import { ObjRef, objRefToString, Identifier, isObjRef } from "../base";
import { idRef } from "../base/factory";
import SparkMD5 from "spark-md5";

type AttributeBuilderInput = Identifier | ObjRef | IAttribute;

/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute}.
 *
 * @public
 */
export class AttributeBuilder {
    private attribute: IAttribute["attribute"];
    private customLocalId: boolean = false;

    /**
     * @internal
     */
    constructor(input: AttributeBuilderInput) {
        if (isAttribute(input)) {
            this.attribute = cloneDeep(input.attribute);
            this.customLocalId = false;
        } else {
            const displayForm: ObjRef = isObjRef(input) ? input : idRef(input, "displayForm");

            this.attribute = {
                displayForm,
                localIdentifier: "",
            };
        }
    }

    public alias = (alias: string) => {
        this.attribute.alias = alias;

        return this;
    };

    public localId = (localId: string) => {
        this.attribute.localIdentifier = localId;
        this.customLocalId = true;

        return this;
    };

    public build = () => {
        const localIdentifier = this.getOrGenerateLocalId();

        return {
            attribute: {
                ...this.attribute,
                localIdentifier,
            },
        };
    };

    private getOrGenerateLocalId(): string {
        if (this.customLocalId && !isEmpty(this.attribute.localIdentifier)) {
            return this.attribute.localIdentifier;
        }

        return ["a", this.calculateAliasHash(), objRefToString(this.attribute.displayForm)]
            .filter(part => !isEmpty(part))
            .join("_");
    }

    private calculateAliasHash(): string {
        if (!this.attribute.alias) {
            return "";
        }

        const hasher = new SparkMD5();
        hasher.append(this.attribute.alias);

        return hasher.end().substr(0, 8);
    }
}

/**
 * Function that will be called to perform modifications of an attribute before it is fully constructed.
 *
 * @public
 */
export type AttributeModifications = (builder: AttributeBuilder) => AttributeBuilder;

/**
 * Creates a new attribute with the specified display form ref and optional modifications and localIdentifier.
 * @param displayFormRefOrId - ref or identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(
    displayFormRefOrId: ObjRef | Identifier,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(displayFormRefOrId);

    return modifications(builder).build();
}

/**
 * Allows modification of an existing attribute instance.
 *
 * @param attribute - attribute to modify
 * @param modifications - modification function
 * @public
 */
export function modifyAttribute(
    attribute: IAttribute,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(attribute);

    return modifications(builder).build();
}
