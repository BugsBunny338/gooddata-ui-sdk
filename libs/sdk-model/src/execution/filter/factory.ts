// (C) 2019-2020 GoodData Corporation
import invariant from "ts-invariant";
import {
    AttributeElements,
    ComparisonConditionOperator,
    IAbsoluteDateFilter,
    IMeasureValueFilter,
    INegativeAttributeFilter,
    IPositiveAttributeFilter,
    IRelativeDateFilter,
    RangeConditionOperator,
} from "./index";
import { IAttribute, attributeAttributeDisplayFormObjRef } from "../attribute";
import { ObjRefInScope, ObjRef, isObjRef, Identifier } from "../base";
import { IMeasure, isMeasure, measureLocalId } from "../measure";
import { idRef } from "../base/factory";

/**
 * Creates a new positive attribute filter.
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param inValues - values to filter for; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newPositiveAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    inValues: AttributeElements | string[],
): IPositiveAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeAttributeDisplayFormObjRef(attributeOrRef);

    const inObject: AttributeElements = Array.isArray(inValues) ? { values: inValues } : inValues;

    return {
        positiveAttributeFilter: {
            displayForm: objRef,
            in: inObject,
        },
    };
}

/**
 * Creates a new negative attribute filter.
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param notInValues - values to filter out; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newNegativeAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    notInValues: AttributeElements | string[],
): INegativeAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeAttributeDisplayFormObjRef(attributeOrRef);

    const notInObject: AttributeElements = Array.isArray(notInValues) ? { values: notInValues } : notInValues;

    return {
        negativeAttributeFilter: {
            displayForm: objRef,
            notIn: notInObject,
        },
    };
}

/**
 * Creates a new absolute date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param from - start of the interval in ISO-8601 calendar date format
 * @param to - end of the interval in ISO-8601 calendar date format
 * @public
 */
export function newAbsoluteDateFilter(
    dateDataSet: ObjRef | Identifier,
    from: string,
    to: string,
): IAbsoluteDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        absoluteDateFilter: {
            dataSet,
            from,
            to,
        },
    };
}

/**
 * Creates a new relative date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param granularity - granularity of the filters (month, year, etc.)
 * @param from - start of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @param to - end of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @public
 */
export function newRelativeDateFilter(
    dateDataSet: ObjRef | Identifier,
    granularity: string,
    from: number,
    to: number,
): IRelativeDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        relativeDateFilter: {
            dataSet,
            granularity,
            from,
            to,
        },
    };
}

/**
 * Creates a new measure value filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - comparison or range operator to use in the filter
 * @param val1 - first numeric value, used as value in comparison or as 'from' value in range operators
 * @param val2 - second numeric value, required in range operators and used in 'to' value; optional and ignored with comparison operators
 * @public
 */
export function newMeasureValueFilter(
    measureOrRef: IMeasure | ObjRefInScope,
    operator: ComparisonConditionOperator | RangeConditionOperator,
    val1: number,
    val2?: number,
): IMeasureValueFilter {
    const ref: ObjRefInScope = isMeasure(measureOrRef)
        ? { localIdentifier: measureLocalId(measureOrRef) }
        : measureOrRef;

    if (operator === "BETWEEN" || operator === "NOT_BETWEEN") {
        invariant(val2, "measure value filter with range operator requires two numeric values");

        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    range: {
                        operator,
                        from: val1,
                        to: val2!,
                    },
                },
            },
        };
    } else {
        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    comparison: {
                        operator,
                        value: val1,
                    },
                },
            },
        };
    }
}
