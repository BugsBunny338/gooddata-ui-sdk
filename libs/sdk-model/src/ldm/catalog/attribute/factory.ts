// (C) 2019-2020 GoodData Corporation
import identity = require("lodash/identity");
import { BuilderModifications, builderFactory } from "../../../base/builder";
import { ICatalogAttribute } from ".";
import { IAttributeMetadataObject } from "../../metadata/attribute";
import { IAttributeDisplayFormMetadataObject } from "../../metadata/attributeDisplayForm";
import { GroupableCatalogItemBuilder } from "../group/factory";
import { newAttributeMetadataObject, AttributeMetadataObjectBuilder } from "../../metadata/attribute/factory";
import { ObjRef, isObjRef } from "../../../objRef";
import {
    AttributeDisplayFormMetadataObjectBuilder,
    newAttributeDisplayFormMetadataObject,
} from "../../metadata";
import { AttributeModifications, newAttribute } from "../../../execution/attribute/factory";
import { IAttribute } from "../../../execution/attribute";

/**
 * @public
 */
export class CatalogAttributeBuilder<
    T extends ICatalogAttribute = ICatalogAttribute
> extends GroupableCatalogItemBuilder<T> {
    public attribute(
        attributeOrRef: IAttributeMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeMetadataObjectBuilder>,
    ): this {
        if (isObjRef(attributeOrRef)) {
            this.item.attribute = newAttributeMetadataObject(attributeOrRef, modifications);
        } else {
            this.item.attribute = attributeOrRef;
        }
        return this;
    }

    public defaultDisplayForm(
        displayFormOrRef: IAttributeDisplayFormMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeDisplayFormMetadataObjectBuilder>,
    ): this {
        if (isObjRef(displayFormOrRef)) {
            this.item.defaultDisplayForm = newAttributeDisplayFormMetadataObject(
                displayFormOrRef,
                modifications,
            );
        } else {
            this.item.defaultDisplayForm = displayFormOrRef;
        }
        return this;
    }

    public toExecutionModel(modifications: AttributeModifications = identity): IAttribute {
        if (!this.item.defaultDisplayForm) {
            throw new Error("Cannot convert catalog attribute to execution model, no displayForm found!");
        }

        const defaultModifications: AttributeModifications = a =>
            a.alias(this.item.defaultDisplayForm?.title!);

        return newAttribute(this.item.defaultDisplayForm.ref, m => modifications(defaultModifications(m)));
    }
}

/**
 * @public
 */
export const newCatalogAttribute = (
    modifications: BuilderModifications<CatalogAttributeBuilder> = identity,
): ICatalogAttribute => builderFactory(CatalogAttributeBuilder, { type: "attribute" }, modifications);

/**
 * @public
 */
export const catalogAttributeToExecutionAttribute = (
    catalogAttribute: ICatalogAttribute,
    modifications: BuilderModifications<CatalogAttributeBuilder> = identity,
): IAttribute => new CatalogAttributeBuilder(catalogAttribute).modify(modifications).toExecutionModel();
