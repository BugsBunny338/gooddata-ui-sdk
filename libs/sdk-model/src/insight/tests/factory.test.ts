// (C) 2020 GoodData Corporation

import { newAttributeSort, newPositiveAttributeFilter } from "../..";
import { Department, Won } from "../../../__mocks__/model";
import { newBucket } from "../../execution/buckets";
import { newInsightDefinition } from "../factory";

describe("insight factory", () => {
    it("should create new empty insight", () => {
        expect(newInsightDefinition("local:bar")).toMatchSnapshot();
    });

    it("should create new empty insight with custom title", () => {
        expect(newInsightDefinition("local:bar", (b) => b.title("My Insight"))).toMatchSnapshot();
    });

    it("should create new empty insight with buckets, filters and sorts", () => {
        expect(
            newInsightDefinition("local:bar", (b) => {
                return b
                    .buckets([newBucket("test", Won)])
                    .filters([newPositiveAttributeFilter(Department, ["Inside Sales"])])
                    .sorts([newAttributeSort(Department)]);
            }),
        ).toMatchSnapshot();
    });

    it("should create new empty insight with custom properties", () => {
        expect(
            newInsightDefinition("local:bar", (b) => b.properties({ controls: { grid: true } })),
        ).toMatchSnapshot();
    });
});
