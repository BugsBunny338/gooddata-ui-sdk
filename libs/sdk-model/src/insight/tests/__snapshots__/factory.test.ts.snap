// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insight factory should create new empty insight 1`] = `
Object {
  "insight": Object {
    "buckets": Array [],
    "filters": Array [],
    "properties": Object {},
    "sorts": Array [],
    "title": "Untitled",
    "visualizationUrl": "local:bar",
  },
}
`;

exports[`insight factory should create new empty insight with buckets, filters and sorts 1`] = `
Object {
  "insight": Object {
    "buckets": Array [
      Object {
        "items": Array [
          Object {
            "measure": Object {
              "definition": Object {
                "measureDefinition": Object {
                  "item": Object {
                    "identifier": "afSEwRwdbMeQ",
                  },
                },
              },
              "localIdentifier": "m_afSEwRwdbMeQ",
            },
          },
        ],
        "localIdentifier": "test",
      },
    ],
    "filters": Array [
      Object {
        "positiveAttributeFilter": Object {
          "displayForm": Object {
            "identifier": "label.owner.department",
            "type": "displayForm",
          },
          "in": Object {
            "values": Array [
              "Inside Sales",
            ],
          },
        },
      },
    ],
    "properties": Object {},
    "sorts": Array [
      Object {
        "attributeSortItem": Object {
          "attributeIdentifier": "a_label.owner.department",
          "direction": "asc",
        },
      },
    ],
    "title": "Untitled",
    "visualizationUrl": "local:bar",
  },
}
`;

exports[`insight factory should create new empty insight with custom properties 1`] = `
Object {
  "insight": Object {
    "buckets": Array [],
    "filters": Array [],
    "properties": Object {
      "controls": Object {
        "grid": true,
      },
    },
    "sorts": Array [],
    "title": "Untitled",
    "visualizationUrl": "local:bar",
  },
}
`;

exports[`insight factory should create new empty insight with custom title 1`] = `
Object {
  "insight": Object {
    "buckets": Array [],
    "filters": Array [],
    "properties": Object {},
    "sorts": Array [],
    "title": "My Insight",
    "visualizationUrl": "local:bar",
  },
}
`;
